### PR TITLE
Scaling ToolStripItem text in PmV2 mode.

### LINKED ISSUE #4872: WinForms HighDPI PerMonitorV2 ToolStrip scales wrong
**.NET Core Version:**

SDK: 5.0.202 (db7cc87d51)
Runtime: 5.0.5 (2f740adc14)

**Problem description:**

A toolstrip with items set to DisplayStyle=ImageAndText does not scale correct.

Starting the sample application provided below on my main screen with 144dpi, shows that both toolbars are of the same height:
image
image

Now moving the window to my secondary screen with 96dpi, suddenly shows that the toolbars are now longer of the same height:
image
image

In my opinion the toolbar only displaying the images has the correct size.

**Minimal repro:**
- Create new .NET-Core 5.0 WinForms app
- Set Application.SetHighDpiMode(HighDpiMode.PerMonitorV2)
- Open the forms designer for Form1
- Set AutoScaleMode = Dpi
- Create 2 ToolBars docked to the top and fill them with standard items
- Edit 1 ToolBar so that the ToolBarItems have DisplayStyle=ImageAndText

I have attached a sample project with the repro steps described above here:
[TestDpiToolStrip.zip](https://github.com/dotnet/winforms/files/6427078/TestDpiToolStrip.zip)

### PR DESCRIPTION
**Pull request by dreddy-work**
Fixes #4872

ToolStripItems are components and have Font property. Components does not receive DPICHANGED messages and ToolstripItems doesn't carry parent-child relationship with owner to get scaled by parent/Container. For these reasons, they need to be explicitly updated for the font when Dpi changes. If font was not explicitly set, DefaultFont, that is scaled to current Dpi, will be applied.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/5059)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.cs [hunk 1]
```diff
@@ -4079,6 +4079,20 @@ protected override void RescaleConstantsForDpi(int deviceDpiOld, int deviceDpiNe
                         _toolStripGrip.ToolStrip_RescaleConstants(deviceDpiOld, deviceDpiNew);
                     }
 
+                    // ToolStripItems are components and have Font property. Components does not
+                    // receive DPICHANGED messages and ToolstripItems doesn't carry parent-child
+                    // relationship with owner to get scaled by parent/Container. For these reasons,
+                    // they need to be explicitly updated for the font when Dpi changes. If font was
+                    // not explicitly set, DefaultFont, that is scaled to current Dpi, will be applied.
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.cs:4086]: We tend to wrap between 120 and 150 char mark:
image

@JeremyKuhne and I use [EditorGuidelines extension](https://marketplace.visualstudio.com/items?itemName=PaulHarrington.EditorGuidelines) that shows the guidelines, and the repo is configured for it.

```suggestion
                    // ToolStripItems are components and have Font property. Components do not receive WM_DPICHANGED messages, nor they have
                    // parent-child relationship with owners and, thus, do not get scaled by parent/Container. For these reasons, they need the font
                    // to be explicitly updated when Dpi changes (only if the font was set explicitly).
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.cs [hunk 2]
```diff
@@ -4079,6 +4079,20 @@ protected override void RescaleConstantsForDpi(int deviceDpiOld, int deviceDpiNe
...
+                    var factor = (float)deviceDpiNew / deviceDpiOld;
+                    foreach (ToolStripItem item in Items)
+                    {
+                        if(item.TryGetExplicitlySetFont(out Font local))
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.cs:4090]: Incorrect spacing
```suggestion
                        if (item.TryGetExplicitlySetFont(out Font local))
```

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Dpi/DpiMessageHelper.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using static Interop;
+
+namespace System.Windows.Forms.Tests.Dpi
+{
+    internal static class DpiMessageHelper
+    {
+        public static IntPtr TriggerDpiMessage(User32.WM message, Control control, int newDpi)
```
[the remark this answers is not in the document]
Comment (dreddy-work) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Dpi/DpiMessageHelper.cs:11]: I am making this a helper method. We may use it in some complex test scenarios. I am just not verifying it in the tests that I wrote.

[the remark this answers is not in the document]
Comment (RussKie) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Dpi/DpiMessageHelper.cs:11]: Please return `void` until there is a use case that requires a return value.

[the remark this answers is not in the document]
Comment (dreddy-work) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Dpi/DpiMessageHelper.cs:11]: Why do you insist? Any benefit?

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Dpi/ToolStripItemTests.Dpi.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            try
+            {
+                int clientWidth = 800;
+                using var form = new Form();
+                form.AutoScaleMode = AutoScaleMode.Dpi;
+                form.ClientSize = new Size(clientWidth, 450);
+                var toolStrip = new ToolStrip();
+                var toolStripItemOpen = new ToolStripButton("Open");
+
+                toolStrip.BackColor = Color.FromArgb(((int)(((byte)(40)))), ((int)(((byte)(40)))), ((int)(((byte)(40)))));
```
[the remark this answers is not in the document]
Comment (dreddy-work) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Dpi/ToolStripItemTests.Dpi.cs:36]: result of barrowing form existing tests :). This test doesn't need back color.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Dpi/DpiMessageHelper.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (RussKie) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Dpi/DpiMessageHelper.cs:11]: Less instructions, faster code.
https://sharplab.io/#v2:C4LglgNgNAJiDUAfAAgJgIwFgBQyDMABGgQMIEDeOB1RhyALAQGoD2YMAFAJQVU0C+fakNoEAkgDtgABWAAnAgCUApsACucid17YaeogHZxU2XIB0ALWVyWAbhGDs/IA

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.cs [hunk 5]
```diff
@@ -4079,6 +4079,18 @@ protected override void RescaleConstantsForDpi(int deviceDpiOld, int deviceDpiNe
                         _toolStripGrip.ToolStrip_RescaleConstants(deviceDpiOld, deviceDpiNew);
                     }
 
+                    // ToolStripItems are components and have Font property. Components do not receive WM_DPICHANGED messages, nor they have
+                    // parent-child relationship with owners and, thus, do not get scaled by parent/Container. For these reasons, they need the font
+                    // to be explicitly updated when Dpi changes (only if the font was set explicitly).
+                    var factor = (float)deviceDpiNew / deviceDpiOld;
+                    foreach (ToolStripItem item in Items)
+                    {
+                        if (item.TryGetExplicitlySetFont(out Font local))
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.cs:4088]: Is this sufficient to scale controls hosted by [ToolsStripControlHost](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.toolstripcontrolhost?view=net-5.0) This is an alternative way to "parent" controls which bypasses the windows SetParent. Each Control has internal ToolStripControlHostProperty and should inherit it's font from that parent.
This probably can be done in a separate PR. Or just open a tracking item for now.

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripItem.cs [hunk 6]
```diff
@@ -882,23 +882,22 @@ public virtual Font Font
...
-                Font f = GetOwnerFont();
-                if (f is not null)
+                font = GetOwnerFont();
+                if (font is not null)
                 {
-                    return f;
+                    return font;
                 }
 
                 return DpiHelper.IsPerMonitorV2Awareness ? _defaultFont : ToolStripManager.DefaultFont;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripItem.cs:896]: I did not find where `_defaultFont` is scaled, could you please point me to it?

**Review by Tanya-Solyanik — DISMISSED** [review]
Looks good, I'm mostly interested in knowing what scenarios had you tested to make sure we don't have double-scaling. And also please follow up on the hosted controls

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Dpi/ToolStripItemTests.Dpi.cs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+            IntPtr originalAwarenessContext = User32.SetThreadDpiAwarenessContext(User32.DPI_AWARENESS_CONTEXT.PER_MONITOR_AWARE_V2);
+
+            try
+            {
+                int clientWidth = 800;
+                using var form = new Form();
+                form.AutoScaleMode = AutoScaleMode.Dpi;
+                form.ClientSize = new Size(clientWidth, 450);
+                var toolStrip = new ToolStrip();
+                var toolStripItemOpen = new ToolStripButton("Open");
```
Comment (RussKie) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Dpi/ToolStripItemTests.Dpi.cs:34]:
```suggestion
                using var toolStrip = new ToolStrip();
                using var toolStripItemOpen = new ToolStripButton("Open");
```

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Dpi/DpiMessageHelper.cs [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+            {
+                User32.WM.DPICHANGED => SendWmDpiChangedMessage(),
+                User32.WM.DPICHANGED_BEFOREPARENT => User32.SendMessageW(control, message, wParam),
+                User32.WM.DPICHANGED_AFTERPARENT => User32.SendMessageW(control, message),
+                _ => throw new NotImplementedException()
+            };
+
+            IntPtr SendWmDpiChangedMessage()
+            {
+                RECT suggestedRect = new(0,
```
[the remark this answers is not in the document]
Comment (dreddy-work) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Dpi/DpiMessageHelper.cs:26]: This is differing from designer but i see this pattern in runtime and keep it consistence.

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.cs (hunk 5 above)
Comment (dreddy-work) [reply src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.cs:4088]: Created a tracking issue. If any of you have a sample app for that scenario, please attach it here: https://github.com/dotnet/winforms/issues/5116

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripItem.cs (hunk 6 above)
Comment (dreddy-work) [reply src/System.Windows.Forms/src/System/Windows/Forms/ToolStripItem.cs:896]: in RescaleConstantsInternal() method.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Dpi/ToolStripItemTests.Dpi.cs [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+{
+    public class ToolStripItemDpiTests : IClassFixture<ThreadExceptionFixture>
+    {
+        [WinFormsTheory]
+        [InlineData(2 * DpiHelper.LogicalDpi)]
+        [InlineData(3.5 * DpiHelper.LogicalDpi)]
+        public void ToolStripItems_FontScaling(int newDpi)
+        {
+            // Run tests only on Windows 10 versions that support thread dpi awareness.
+            if (!PlatformDetection.IsWindows10Version1803OrGreater)
```
[the remark this answers is not in the document]
Comment (dreddy-work) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Dpi/ToolStripItemTests.Dpi.cs:19]: Given that the version of windows 10 that is still in support must have the thread DpiAwareness.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Dpi/ToolStripItemTests.Dpi.cs [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+                toolStrip.Name = "toolStrip1";
+                toolStrip.Text = "toolStrip1";
+                using Font initialFont = toolStrip.Font = new Font("Segoe UI", 9F, FontStyle.Bold, GraphicsUnit.Point);
+
+                form.Show();
+
+                DpiMessageHelper.TriggerDpiMessage(User32.WM.DPICHANGED_BEFOREPARENT, toolStrip, newDpi);
+                var factor = newDpi / DpiHelper.LogicalDpi;
+
+                Assert.Equal((float)initialFont.Size * factor, toolStrip.Font.Size);
```
[the remark this answers is not in the document]
Comment (dreddy-work) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Dpi/ToolStripItemTests.Dpi.cs:48]: I see Item size is not impacted by Font change here.

**Comment by dreddy-work** [comment]
> Looks good, I'm mostly interested in knowing what scenarios had you tested to make sure we don't have double-scaling. And also please follow up on the hosted controls

Created a tracking bug on HostedControl scenario. I did a manual verification with test application with derived Font and Explicitly set Font. Scaling look fine here. CTI will be doing general testing.

**Comment by dreddy-work** [comment]
@Tanya-Solyanik / @RussKie , Any further feedback or good to go?

**Review by RussKie — APPROVED** [review]

**Pull request merged**